File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/* 
 For every root they are three paths the maxStraight path  or the max of the all the node with the root 
 To reach the children first we need to use dfs in the left right root order such that we calculate all
 the children first then the root 
*/
class Solution {
    int ans = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        helper(root);
        return ans;
        }
    int helper(TreeNode* root){
        if(root == nullptr) return 0;
      
        int x = max(0,helper(root->left));
        int y = max(0,helper(root->right));
        ans = max(ans, root->val + x + y);


    return max(x,y) + root->val;
    }    
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxAns;

    public int maxPathSum(TreeNode root) {
        maxAns = Integer.MIN_VALUE;
        dfs(root);
        return maxAns;
    }

    int dfs(TreeNode root) {
        if (root == null)
            return 0;
        int maxRight = dfs(root.right);
        int maxLeft = dfs(root.left);
        int currmax = Math.max(root.val + maxRight, root.val + maxLeft);
        currmax = Math.max(root.val, currmax);
        maxAns = Math.max(maxLeft + maxRight + root.val, maxAns);
        maxAns = Math.max(maxAns, currmax);
        

        return currmax;
    }
}